v3/lints/rfc/lint_name_constraint_not_fqdn.go

```diff
@@ -0,0 +1,118 @@
+/*
+ * ZLint Copyright 2021 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package rfc
+
+import (
+	"strings"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+)
+
+type nameConstraintNotFQDN struct{}
+
+func (l *nameConstraintNotFQDN) Initialize() error {
+	return nil
+}
+
+func (l *nameConstraintNotFQDN) CheckApplies(c *x509.Certificate) bool {
+	return util.IsExtInCert(c, util.NameConstOID)
+}
+
+func (l *nameConstraintNotFQDN) Execute(c *x509.Certificate) *lint.LintResult {
+
+	incorrectPermittedHosts := make([]string, 0)
+	incorrectExcludedHosts := make([]string, 0)
+	errString := ""
+
+	for _, subtreeString := range c.PermittedURIs {
+		isHostFQDN(subtreeString.Data, &incorrectPermittedHosts)
+	}
+	for _, subtreeString := range c.ExcludedURIs {
+		isHostFQDN(subtreeString.Data, &incorrectExcludedHosts)
+	}
+
+	if len(incorrectPermittedHosts) != 0 {
+		errString += returnErrorString(incorrectPermittedHosts, true)
+	}
+	if len(incorrectPermittedHosts) != 0 && len(incorrectExcludedHosts) != 0 {
+		errString += "; "
+	}
+	if len(incorrectExcludedHosts) != 0 {
+		errString += returnErrorString(incorrectExcludedHosts, false)
+	}
+
+	if len(errString) != 0 {
+		return &lint.LintResult{
+			Status:  lint.Error,
+			Details: errString,
+		}
+	}
+
+	return &lint.LintResult{Status: lint.Pass}
+}
+
+func isHostFQDN(host string, incorrectHosts *[]string) {
```

Comment:
In the effort of keeping functions doing one job only I think it would be a little bit more natural to write this as a function that returns a `bool` for the `host` and doesn't do the job of collecting up the `incorrectHosts` (the caller would do that). That has a few advantages:

1. When folks see a function named "isXXX" the natural expectation is that it returns true/false.
1. It avoids the need to pass the `incorrectHosts` slice in by reference so that this function can mutate it on L73. It's harder to reason about code that shares a var like this.

If you wanted to keep things mostly the way you have them I'd change `isHostFQDN` to be something like:

```go
func isHostFQDN(host string) bool {
   host = strings.TrimPrefix(host, ".")
   return util.IsFQDN(host)
}
```

Alternatively, what I was originally thinking (sorry if I didn't make this very clear) was you could write a function that takes a `[]string` as argument and returns a `[]string` of all of the entries that are **not** an FQDN. Then the caller only has to check if the len of the returned `[]string` is non-zero and if it is, process into a failure result & message.